File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.cc
#include<iostream>
#include<queue>
using namespace std;
int arr[1005][1005],c[1005][1005],visit[1005][1005];
int n,m;

int dx[4] = {0,0,-1,1};
int dy[4] = {1,-1,0,0};

bool inrange(int x, int y){
    return x >=0 && x < n && y >=0 && y < m;
}
queue<pair<int,int>>q;

int main()
{
    ios::sync_with_stdio(0); cin.tie(0);
    cin >> n >> m;
    for(int i = 0; i<n; i++){
        for(int j = 0; j<m; j++) {
            cin >> arr[i][j];
            if(arr[i][j] == 2) q.push({i,j});
        }
    }
    
    while(!q.empty()){
        int cx = q.front().first;
        int cy = q.front().second;
        q.pop();
        if(visit[cx][cy]) continue;
        visit[cx][cy] = 1;
        for(int i = 0; i<4; i++){
            int nx = cx + dx[i];
            int ny = cy + dy[i];
            if(!inrange(nx,ny) || !arr[nx][ny] || visit[nx][ny]) continue;
            c[nx][ny] = c[cx][cy] + 1;
            q.push({nx,ny});
        }
    }
    
    for(int i = 0; i<n; i++){
        for(int j = 0; j<m; j++) {
            if(arr[i][j]==1 && !c[i][j]) cout << -1 << ' ';
            else cout << c[i][j] << ' ';
        }
        cout << '\n';
    }
    
    return 0;
}

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.java
import java.io.*;
import java.util.*;

class Main {
	public static int n,m,sx,sy;
	public static int[][] arr;
	public static int[] dx = {-1,1,0,0};
	public static int[] dy = {0,0,-1,1};
	public static boolean inrange(int x, int y){
		return x >= 0 && x < n && y >= 0 && y < m;
	}
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		
		int[][] arr = new int[n][m];
		int[][] visit = new int[n][m];
		
		
		for(int i = 0; i < n; i++){
			StringTokenizer st2 = new StringTokenizer(br.readLine());
			for(int j = 0; j < m; j++){
				arr[i][j]= Integer.parseInt(st2.nextToken());
				if(arr[i][j] == 2) {
					sx = i;
					sy = j;
					visit[sx][sy] = 1;
				}
			}
		}
		
		Queue<Pair>queue = new LinkedList<>();
		queue.offer(new Pair(sx,sy,0));
		
		while(queue.size() > 0){
			Pair now = queue.poll();
			int cx = now.x;
			int cy = now.y;
			int d = now.d;
			
			for(int i = 0; i < 4; i++){
				int nx = cx + dx[i];
				int ny = cy + dy[i];
				if(!inrange(nx,ny)) continue;
				if(visit[nx][ny] != 0) continue;
				if(arr[nx][ny] == 0) continue;
				visit[nx][ny] = d + 1;
				queue.offer(new Pair(nx,ny,d+1));
			}
		}
		
		StringBuilder sb = new StringBuilder();
		visit[sx][sy] = 0;
		for(int i = 0; i<n; i++){
			for(int j = 0; j<m; j++){
				if(visit[i][j] == 0 && arr[i][j] == 1){
					sb.append(-1 + " ");
				}
				else sb.append(visit[i][j] + " ");
			}
			sb.append("\n");
		}
		
		System.out.println(sb.toString());
	}
	
	static class Pair{
		int x;
		int y;
		int d;
		
		public Pair(int x, int y, int d){
			this.x = x;
			this.y = y;
			this.d = d;
		}
	}
}